Add Check Support for Java 21 Pattern Matching for Switch Syntax: SimplifyBooleanReturn

child of https://github.com/checkstyle/checkstyle/issues/14961

I have read check documentation:https://checkstyle.org/checks/coding/simplifybooleanreturn.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(boolean p, int x, String c) { }
record Rectangle(boolean upperLeft, boolean lowerRight) { }
public class Test {

    boolean test1(Object o) {
       if (o instanceof ColoredPoint(boolean x, _, _)) {
            if (x) return true;   // violation
            else return false;
        } else {
            return false;
        }
    }
    boolean test2(Object obj) {
        return switch (obj) {
            case ColoredPoint(boolean x,_,_) -> {
                if (x) yield true;       // expected violation , can be yield x
                else yield false;
            }
            default -> false;
        };
    }
}
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="SimplifyBooleanReturn">

        </module>
    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:7:9: Conditional logic can be removed. [SimplifyBooleanReturn]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 
```

Describe what you want in detail

I am expecting a violation on unnecessary conditional logic that can be simplified inside switch expressions. each case block is an execution path for a return. So this condition falls under a Boolean return

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanReturnCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanReturnCheck.java
@@ -27,7 +27,7 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 /**
  * <p>
- * Checks for over-complicated boolean return statements.
+ * Checks for over-complicated boolean return or yield statements.
  * For example the following code
  * </p>
  * <pre>
@@ -105,55 +105,56 @@ public class SimplifyBooleanReturnCheck
             final DetailAST condition = ast.getFirstChild().getNextSibling();
             final DetailAST thenStatement = condition.getNextSibling().getNextSibling();
 
-            if (canReturnOnlyBooleanLiteral(thenStatement)
-                && canReturnOnlyBooleanLiteral(elseStatement)) {
+            if (canReturnOrYieldOnlyBooleanLiteral(thenStatement)
+                && canReturnOrYieldOnlyBooleanLiteral(elseStatement)) {
                 log(ast, MSG_KEY);
             }
         }
     }
 
     /**
-     * Returns if an AST is a return statement with a boolean literal
-     * or a compound statement that contains only such a return statement.
+     * Returns if an AST is a return or a yield statement with a boolean literal
+     * or a compound statement that contains only such a return or a yield statement.
      *
      * <p>Returns {@code true} iff ast represents
      * <pre>
-     * return true/false;
+     * return/yield true/false;
      * </pre>
      * or
      * <pre>
      * {
-     *   return true/false;
+     *   return/yield true/false;
      * }
      * </pre>
      *
      * @param ast the syntax tree to check
-     * @return if ast is a return statement with a boolean literal.
+     * @return if ast is a return or a yield statement with a boolean literal.
      */
-    private static boolean canReturnOnlyBooleanLiteral(DetailAST ast) {
+    private static boolean canReturnOrYieldOnlyBooleanLiteral(DetailAST ast) {
         boolean result = true;
-        if (!isBooleanLiteralReturnStatement(ast)) {
+        if (!isBooleanLiteralReturnOrYieldStatement(ast)) {
             final DetailAST firstStatement = ast.getFirstChild();
-            result = isBooleanLiteralReturnStatement(firstStatement);
+            result = isBooleanLiteralReturnOrYieldStatement(firstStatement);
         }
         return result;
     }
 
     /**
-     * Returns if an AST is a return statement with a boolean literal.
+     * Returns if an AST is a return or a yield statement with a boolean literal.
      *
      * <p>Returns {@code true} iff ast represents
      * <pre>
-     * return true/false;
+     * return/yield true/false;
      * </pre>
      *
      * @param ast the syntax tree to check
-     * @return if ast is a return statement with a boolean literal.
+     * @return if ast is a return or a yield statement with a boolean literal.
      */
-    private static boolean isBooleanLiteralReturnStatement(DetailAST ast) {
+    private static boolean isBooleanLiteralReturnOrYieldStatement(DetailAST ast) {
         boolean booleanReturnStatement = false;
 
-        if (ast != null && ast.getType() == TokenTypes.LITERAL_RETURN) {
+        if (ast != null && (ast.getType() == TokenTypes.LITERAL_RETURN
+                                || ast.getType() == TokenTypes.LITERAL_YIELD)) {
             final DetailAST expr = ast.getFirstChild();
 
             if (expr.getType() != TokenTypes.SEMI) {

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/SimplifyBooleanReturnCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/SimplifyBooleanReturnCheck.xml
@@ -5,7 +5,7 @@
              name="SimplifyBooleanReturn"
              parent="com.puppycrawl.tools.checkstyle.TreeWalker">
          <description>&lt;p&gt;
- Checks for over-complicated boolean return statements.
+ Checks for over-complicated boolean return or yield statements.
  For example the following code
  &lt;/p&gt;
  &lt;pre&gt;

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanReturnCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/SimplifyBooleanReturnCheckTest.java
@@ -45,6 +45,18 @@ public class SimplifyBooleanReturnCheckTest
     }
 
     @Test
+    public void testYield() throws Exception {
+        final String[] expected = {
+            "14:17: " + getCheckMessage(MSG_KEY),
+            "28:17: " + getCheckMessage(MSG_KEY),
+            "39:17: " + getCheckMessage(MSG_KEY),
+            "54:17: " + getCheckMessage(MSG_KEY),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputSimplifyBooleanReturnWithYield.java"), expected);
+    }
+
+    @Test
     public void testTokensNotNull() {
         final SimplifyBooleanReturnCheck check = new SimplifyBooleanReturnCheck();
         assertWithMessage("Acceptable tokens should not be null")

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/simplifybooleanreturn/InputSimplifyBooleanReturnWithYield.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/simplifybooleanreturn/InputSimplifyBooleanReturnWithYield.java
@@ -1,0 +1,89 @@
+/*
+SimplifyBooleanReturn
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.simplifybooleanreturn;
+
+public class InputSimplifyBooleanReturnWithYield {
+
+    boolean test(int x, boolean c) {
+        return switch (x) {
+            case 1 -> {
+                if (c) { // violation
+                    yield true;
+                } else {
+                    yield false;
+                }
+            }
+            case 2 -> false;
+            default -> false;
+        };
+    }
+
+    boolean test2(int x, boolean c) {
+        return switch (x) {
+            case 1 -> {
+                if (c) yield true; // violation
+                else yield false;
+            }
+            case 2 -> false;
+            default -> false;
+        };
+    }
+
+    boolean test3(int x, boolean c) {
+        return switch (x) {
+            case 1: {
+                if (c) yield true; // violation
+                else yield false;
+            }
+            case 2: {
+                yield false;
+            }
+            default: {
+                yield false;
+            }
+        };
+    }
+
+    boolean test4(int x, boolean c) {
+        switch (x) {
+            case 1: {
+                if (c) { // violation
+                    return true;
+                } else {
+                    return false;
+                }
+            }
+            case 2: {
+                return false;
+            }
+            default: {
+                return false;
+            }
+        }
+    }
+
+    boolean correct(int x, boolean c) {
+        boolean y = switch (x) {
+            case 1: {
+                yield c;
+            }
+            case 2: {
+                yield false;
+            }
+            default: {
+                yield false;
+            }
+        };
+        return switch (x) {
+            case 1 -> {
+                yield c;
+            }
+            case 2 -> false;
+            default -> false;
+        };
+    }
+}

--- a/src/xdocs/checks.xml
+++ b/src/xdocs/checks.xml
@@ -1204,7 +1204,7 @@
                 SimplifyBooleanReturn
               </a>
             </td>
-            <td>Checks for over-complicated boolean return statements.</td>
+            <td>Checks for over-complicated boolean return or yield statements.</td>
           </tr>
           <tr>
             <td>

--- a/src/xdocs/checks/coding/index.xml
+++ b/src/xdocs/checks/coding/index.xml
@@ -478,7 +478,7 @@
               </a>
             </td>
             <td>
-              Checks for over-complicated boolean return statements.
+              Checks for over-complicated boolean return or yield statements.
             </td>
           </tr>
           <tr>

--- a/src/xdocs/checks/coding/simplifybooleanreturn.xml
+++ b/src/xdocs/checks/coding/simplifybooleanreturn.xml
@@ -10,7 +10,7 @@
       <p>Since Checkstyle 3.0</p>
       <subsection name="Description" id="Description">
         <p>
-          Checks for over-complicated boolean return statements. For example
+          Checks for over-complicated boolean return or yield statements. For example
           the following code
         </p>
         <source>

--- a/src/xdocs/checks/coding/simplifybooleanreturn.xml.template
+++ b/src/xdocs/checks/coding/simplifybooleanreturn.xml.template
@@ -10,7 +10,7 @@
       <p>Since Checkstyle 3.0</p>
       <subsection name="Description" id="Description">
         <p>
-          Checks for over-complicated boolean return statements. For example
+          Checks for over-complicated boolean return or yield statements. For example
           the following code
         </p>
         <source>

